packages/opentelemetry-basic-tracer/src/Span.ts

```diff
@@ -73,6 +76,19 @@ export class Span implements types.Span, ReadableSpan {
 
   setAttribute(key: string, value: unknown): this {
     if (this._isSpanEnded()) return this;
+
+    if (
+      Object.keys(this.attributes).length >=
+      this._traceParams.numberOfAttributesPerSpan!
```

Comment:
When might `nubmerOfAttributesPerSpan` be undefined? Can we make sure that it's not undefined in the constructor, e.g. by using either what's given or a default value? I think that's a safer than doing a force unwrap here as otherwise this will always be false

Same goes for other checks below

Reply:
All the properties in `TraceParam` are always available at this stage. We have a [merge config](https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-basic-tracer/src/BasicTracer.ts#L58) function in BasicTracer which assigns default value, if not provided by the user.

Reply:
OK, I guess what I'm saying is should we have a separate type for when all the params are available, e.g. something like `FullTraceParams` with no optional values? Then when we merge the config, we internally store it with that type. That makes the type system better communicate what is going on.

Reply:
Hmm, I see your point. Will address that later in separate PR.